File: 826.py
import bisect
import collections
from typing import List

class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        n = len(difficulty)
        best_work = collections.defaultdict(int)
        for d, p in zip(difficulty, profit):
            best_work[d] = max(p, best_work[d])

        difficulty.sort()
        for i in range(1, n):
            best_work[difficulty[i]] = max(best_work[difficulty[i-1]], best_work[difficulty[i]])
        # print(best_work)
        # print(difficulty)
        ans = 0
        idx = 0
        pre_idx = 0
        for w in sorted(worker):
            idx = bisect.bisect_left(difficulty, w, pre_idx)
            if idx == n or difficulty[idx] > w:
                idx -= 1
            if idx < 0:
                continue
            pre_idx = idx
            # print(w, idx, difficulty[idx])
            ans += best_work[difficulty[idx]]
        return ans

if __name__ == '__main__':
    s = Solution()
    difficulty = [68,35,52,47,86]
    profit = [67,17,1,81,3]
    worker = [92,10,85,84,82]
    print(s.maxProfitAssignment(difficulty, profit, worker))

File: 826.go
package leetcode

import "sort"

type Job struct {
	difficulty int
	profit     int
}

func maxProfitAssignment(difficulty []int, profit []int, worker []int) int {
	n := len(difficulty)
	jobs := make([]Job, n)
	for i, d := range difficulty {
		jobs[i] = Job{d, profit[i]}
	}
	sort.Slice(jobs, func(i, j int) bool {
		return jobs[i].difficulty < jobs[j].difficulty
	})
	ans := 0
	sort.Ints(worker)
	idx := 0
	maxProfit := 0
	for _, skill := range worker {
		for idx < n && jobs[idx].difficulty <= skill {
			if jobs[idx].profit > maxProfit {
				maxProfit = jobs[idx].profit
			}
			idx++
		}
		ans += maxProfit
	}
	return ans
}
